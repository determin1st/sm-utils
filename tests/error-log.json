[
    {
        "level": 0,
        "msg": [
            "A major change was made by the first Emperor,\nAugustus (27 BC — 14 AD), who reformed revenue\ncollection, bringing large parts of Rome’s empire\nunder consistent direct taxation instead of asking\nfor intermittent tributes. Taxation was determined\nby population census and private tax farming\nwas abolished in favor for civil service tax collectors.\nThis made provincial administration far more tolerable,\ndecreased corruption and oppression and\nincreased revenues."
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 2,
        "msg": [
            "Slavery burden",
            "\nWhile it is true that the Roman Empire wasn’t as heavily\nbureaucratized as the Han Dynasty of Imperial China,\nthe Roman Empire did have a bureaucracy even in this early\nimperial period that tends to be underestimated; the officials,\nas stated above, had a large number of slaves who worked\ninformally in administrative jobs."
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 0,
        "msg": [
            "history",
            "city-states",
            "confederation\nRome has been called by some historians a confederation\nof city-states and that is true as the Roman state\npreferred on a local level to delegate tasks to city-states\nand in provinces that did not have such traditions,\nRome brought them into being as far as it could.\nCity authorities had to maintain order and extract revenue\nnot only for the city itself but also from the countryside\nthat was allocated to that city and where\nthe majority of the population lived."
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 2,
        "msg": [
            "Empire",
            "problems",
            "Tax burden",
            "\nUnder Diocletian and Constantine in the late third and\nearly fourth centuries, the Roman Empire underwent an even\ngreater bureaucratization in order to deal with the threat\nof external enemies and internal instability.\nThe Empire was to be ruled by two Emperors, one in the West\nand one in the East, who could respond faster to internal\nand external threats. In order to maintain a large army\nto defend the Empire from the ‘barbarian’ threat,\nmore taxes were needed."
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 0,
        "msg": [
            "one"
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 0,
        "msg": [
            "one",
            "two"
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 0,
        "msg": [
            "one",
            "two",
            "three"
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": []
    },
    {
        "level": 2,
        "msg": [
            "go",
            "deeper"
        ],
        "span": 123456,
        "time": 1700073135,
        "logs": [
            {
                "level": 2,
                "msg": [
                    "one"
                ]
            },
            {
                "level": 0,
                "msg": [
                    "one",
                    "two"
                ]
            },
            {
                "level": 0,
                "msg": [
                    "one",
                    "two",
                    "three"
                ]
            },
            {
                "level": 0,
                "msg": [
                    "one",
                    "two",
                    "three",
                    "four"
                ]
            },
            {
                "level": 0,
                "msg": [
                    "one",
                    "two",
                    "three",
                    "four",
                    "five"
                ]
            },
            {
                "level": 0,
                "msg": [
                    "one two three"
                ],
                "logs": [
                    {
                        "level": 1,
                        "msg": [
                            "one",
                            "two",
                            "three",
                            "four",
                            "five\nthe bunny went for a walk\nsuddenly, the hunter runs out and.."
                        ],
                        "logs": [
                            {
                                "level": 1,
                                "msg": [
                                    "..shots stright away!\noh my, oh my"
                                ],
                                "logs": [
                                    {
                                        "level": 1,
                                        "msg": [
                                            "message 1\nmessage 2\nmessage 3"
                                        ]
                                    },
                                    {
                                        "level": 1,
                                        "msg": [
                                            "bunny is dead"
                                        ]
                                    },
                                    {
                                        "level": 1,
                                        "msg": [
                                            "bunny\nis dead"
                                        ]
                                    },
                                    {
                                        "level": 1,
                                        "msg": [
                                            "bunny",
                                            "is\ndead"
                                        ]
                                    }
                                ]
                            },
                            {
                                "level": 0,
                                "msg": [
                                    "everything is fine"
                                ]
                            }
                        ]
                    },
                    {
                        "level": 0,
                        "msg": [
                            "one",
                            "two",
                            "three",
                            "four",
                            "five"
                        ]
                    },
                    {
                        "level": 0,
                        "msg": [
                            "one",
                            "two",
                            "three",
                            "four"
                        ]
                    },
                    {
                        "level": 0,
                        "msg": [
                            "one",
                            "two",
                            "three"
                        ]
                    },
                    {
                        "level": 0,
                        "msg": [
                            "one",
                            "two"
                        ]
                    }
                ]
            }
        ]
    }
]