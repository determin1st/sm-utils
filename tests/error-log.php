<?php declare(strict_types=1);
require_once
  __DIR__.DIRECTORY_SEPARATOR.
  '..'.DIRECTORY_SEPARATOR.
  'conio.php';
###
$logs = [
# top multilines {{{
[
  'level' => 0,
  'msg'   => [# type=2 (no title)
    "A major change was made by the first Emperor,\n".
    "Augustus (27 BC — 14 AD), who reformed revenue\n".
    "collection, bringing large parts of Rome’s empire\n".
    "under consistent direct taxation instead of asking\n".
    "for intermittent tributes. Taxation was determined\n".
    "by population census and private tax farming\n".
    "was abolished in favor for civil service tax collectors.\n".
    "This made provincial administration far more tolerable,\n".
    "decreased corruption and oppression and\n".
    "increased revenues."
  ],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
[
  'level' => 2,
  'msg'   => [# type=2 (title)
    'Slavery burden',
    "\n".
    "While it is true that the Roman Empire wasn’t as heavily\n".
    "bureaucratized as the Han Dynasty of Imperial China,\n".
    "the Roman Empire did have a bureaucracy even in this early\n".
    "imperial period that tends to be underestimated; the officials,\n".
    "as stated above, had a large number of slaves who worked\n".
    "informally in administrative jobs."
  ],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
[
  'level' => 0,
  'msg'   => [# type=3 (no title)
    'history','city-states',"confederation\n".
    "Rome has been called by some historians a confederation\n".
    "of city-states and that is true as the Roman state\n".
    "preferred on a local level to delegate tasks to city-states\n".
    "and in provinces that did not have such traditions,\n".
    "Rome brought them into being as far as it could.\n".
    "City authorities had to maintain order and extract revenue\n".
    "not only for the city itself but also from the countryside\n".
    "that was allocated to that city and where\n".
    "the majority of the population lived."
  ],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
[
  'level' => 2,
  'msg'   => [# type=3 (title)
    'Empire','problems','Tax burden',
    "\n".
    "Under Diocletian and Constantine in the late third and\n".
    "early fourth centuries, the Roman Empire underwent an even\n".
    "greater bureaucratization in order to deal with the threat\n".
    "of external enemies and internal instability.\n".
    "The Empire was to be ruled by two Emperors, one in the West\n".
    "and one in the East, who could respond faster to internal\n".
    "and external threats. In order to maintain a large army\n".
    "to defend the Empire from the ‘barbarian’ threat,\n".
    "more taxes were needed."
  ],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
# }}}
# oneliners {{{
[
  'level' => 0,
  'msg'   => ['one'],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
[
  'level' => 0,
  'msg'   => ['one','two'],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
[
  'level' => 0,
  'msg'   => ['one','two','three'],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  => [],
],
# }}}
# nesting {{{
[
  'level' => 2,
  'msg'   => ['go','deeper'],
  'span'  => 123456,
  'time'  => 1700073135,
  'logs'  =>
  [
    [
      'level' => 2,
      'msg'   => ['one'],
    ],
    [
      'level' => 0,
      'msg'   => ['one','two'],
    ],
    [
      'level' => 0,
      'msg'   => ['one','two','three'],
    ],
    [
      'level' => 0,
      'msg'   => ['one','two','three','four'],
    ],
    [
      'level' => 0,
      'msg'   => ['one','two','three','four','five'],
    ],
    ###
    [
      'level' => 0,
      'msg'   => ['one two three'],
      'logs'  =>
      [
        [
          'level' => 1,
          'msg'   => [
            'one','two','three','four',
            "five\n".
            "the bunny went for a walk\n".
            "suddenly, the hunter runs out and.."
          ],
          'logs'  =>
          [
            [
              'level' => 1,
              'msg'   => [
                "..shots stright away!\n".
                "oh my, oh my"
              ],
              'logs'  =>
              [
                [
                  'level' => 1,
                  'msg'   => ["message 1\nmessage 2\nmessage 3"],
                ],
                [
                  'level' => 1,
                  'msg'   => ['bunny is dead'],
                ],
                [
                  'level' => 1,
                  'msg'   => ["bunny\nis dead"],
                ],
                [
                  'level' => 1,
                  'msg'   => ["bunny","is\ndead"],
                ],
              ],
            ],
            [
              'level' => 0,
              'msg'   => ['everything is fine'],
            ],
          ],
        ],
        ###
        [
          'level' => 0,
          'msg'   => ['one','two','three','four','five'],
        ],
        [
          'level' => 0,
          'msg'   => ['one','two','three','four'],
        ],
        [
          'level' => 0,
          'msg'   => ['one','two','three'],
        ],
        [
          'level' => 0,
          'msg'   => ['one','two'],
        ],
      ],
    ],
  ],
],
# }}}
];
if (0)
{
  file_put_contents(
    substr(__FILE__, 0, -4).'.json',
    json_encode($logs,  JSON_PRETTY_PRINT|JSON_UNESCAPED_UNICODE)
  );
}
if (class_exists('SM\Conio', false))
{
  \SM\ErrorLog::init([
    'ansi' => \SM\Conio::is_ansi()
  ]);
}
$out = \SM\ErrorLog::render($logs);
echo $out;
if (0)
{
  file_put_contents(
    substr(__FILE__, 0, -4).'.out',
    #mb_convert_encoding($out, 'UTF-16BE', 'UTF-8')
    #iconv('UTF-8', 'CP866', $out)
    mb_convert_encoding($out, 'CP866', 'UTF-8')
    #$out
  );
}
###
