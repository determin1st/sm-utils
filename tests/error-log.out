þ2023/11/15 23:32:15 PASS
³Ö
À¶ A major change was made by the first Emperor,
 º Augustus (27 BC ? 14 AD), who reformed revenue
 º collection, bringing large parts of Rome?s empire
 º under consistent direct taxation instead of asking
 º for intermittent tributes. Taxation was determined
 º by population census and private tax farming
 º was abolished in favor for civil service tax collectors.
 º This made provincial administration far more tolerable,
 º decreased corruption and oppression and
 º increased revenues.
 Ó ~123úus
þ2023/11/15 23:32:15 FAIL
³Ö
À¶ Slavery burden
 º While it is true that the Roman Empire wasn?t as heavily
 º bureaucratized as the Han Dynasty of Imperial China,
 º the Roman Empire did have a bureaucracy even in this early
 º imperial period that tends to be underestimated; the officials,
 º as stated above, had a large number of slaves who worked
 º informally in administrative jobs.
 Ó ~123úus
þ2023/11/15 23:32:15 PASS: history > city-states > confederation ~123úus
³Ö
À¶ Rome has been called by some historians a confederation
 º of city-states and that is true as the Roman state
 º preferred on a local level to delegate tasks to city-states
 º and in provinces that did not have such traditions,
 º Rome brought them into being as far as it could.
 º City authorities had to maintain order and extract revenue
 º not only for the city itself but also from the countryside
 º that was allocated to that city and where
 º the majority of the population lived.
 Ó
þ2023/11/15 23:32:15 FAIL: Empire > problems ~123úus
³Ö
À¶ Tax burden
 º Under Diocletian and Constantine in the late third and
 º early fourth centuries, the Roman Empire underwent an even
 º greater bureaucratization in order to deal with the threat
 º of external enemies and internal instability.
 º The Empire was to be ruled by two Emperors, one in the West
 º and one in the East, who could respond faster to internal
 º and external threats. In order to maintain a large army
 º to defend the Empire from the ?barbarian? threat,
 º more taxes were needed.
 Ó
þ2023/11/15 23:32:15 PASS: one ~123úus
þ2023/11/15 23:32:15 PASS: one > two ~123úus
þ2023/11/15 23:32:15 PASS: one > two > three ~123úus
þ2023/11/15 23:32:15 FAIL: go > deeper ~123úus
Ã>ERROR: one
Ã>INFO: one > two
Ã>INFO: oneútwo > three
Ã>INFO: oneútwoúthree > four
Ã>INFO: oneútwoúthreeúfour > five
À>INFO: one two three
 Ã>WARNING: oneútwoúthreeúfour > five
 ³³Ö
 ³Ã¶ the bunny went for a walk
 ³³º suddenly, the hunter runs out and..
 ³³Ó
 ³Ã>WARNING
 ³³³Ö
 ³³Ã¶ ..shots stright away!
 ³³³º oh my, oh my
 ³³³Ó
 ³³Ã>WARNING
 ³³³³Ö
 ³³³À¶ message 1
 ³³³ º message 2
 ³³³ º message 3
 ³³³ Ó
 ³³Ã>WARNING: bunny is dead
 ³³Ã>WARNING
 ³³³³Ö
 ³³³À¶ bunny
 ³³³ º is dead
 ³³³ Ó
 ³³À>WARNING: bunny > is
 ³³ ³Ö
 ³³ À¶ dead
 ³³  Ó
 ³À>INFO: everything is fine
 Ã>INFO: oneútwoúthreeúfour > five
 Ã>INFO: oneútwoúthree > four
 Ã>INFO: oneútwo > three
 À>INFO: one > two
